orttraining/orttraining/test/python/orttraining_test_load_checkpoint.py

```diff
@@ -0,0 +1,1475 @@
+#!/usr/bin/env python3
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import os
+import pickle
+from numpy.testing import assert_allclose
+import numpy as np
+import argparse
+import glob
+
+import torch
+import torch.distributed as dist
+
+import onnxruntime
+from onnxruntime import set_seed
+from onnxruntime.training import amp, checkpoint, optim, orttrainer
+from onnxruntime.capi._pybind_state import set_cuda_device_id, get_mpi_context_world_rank, get_mpi_context_world_size
+from orttraining_test_orttrainer_frontend import _load_pytorch_transformer_model
+
+def distributed_setup(save_function):
+    def setup():
+        world_rank = get_mpi_context_world_rank()
+        world_size = get_mpi_context_world_size()
+        device = 'cuda:' + str(world_rank)
+
+        os.environ['RANK'] = str(world_rank)
+        os.environ['WORLD_SIZE'] = str(world_size)
+        os.environ['MASTER_ADDR'] = '127.0.0.1'
+        os.environ['MASTER_PORT'] = '29500'
+
+        set_cuda_device_id(world_rank)
+
+        dist.init_process_group(backend='nccl', world_size=world_size, rank=world_rank)
+        save_function(world_rank, world_size, device)
+    return setup
+
+def load_evaluate_extract_state_dict(device, trainer_opts, checkpoint_dir):
+    seed = 1
+    torch.manual_seed(seed)
+    set_seed(seed)
+
+    # PyTorch transformer model setup
+    learning_rate = 0.1
+    optim_config = optim.LambConfig(lr=learning_rate)
+    model, model_desc, loss_fn, batcher_fn, train_data, _, _ = _load_pytorch_transformer_model(device)
+    trainer = orttrainer.ORTTrainer(model, model_desc, optim_config, loss_fn=loss_fn, options=trainer_opts)
+
+    # load checkpoint into trainer
+    checkpoint.experimental_load_checkpoint(trainer, checkpoint_dir)
+
+    # run an eval step to innitialize the graph
+    torch.manual_seed(seed)
+    set_seed(seed)
+    data, targets = batcher_fn(train_data, 0)
+    trainer.eval_step(data, targets)
+
+    return checkpoint.experimental_state_dict(trainer), model
+
+def split_state_dict(state_dict):
```

Comment:
There are some scenarios that cannot be tested directly; for example going from a single node run to a distributed run. Workflow that illustrates why this scenario cannot be tested directly.
- Two files are saved through ```orttraining_test_save_checkpoint.py```
  - state dictionary through pickle after being extracted through ```experimental_state_dict``` method
  - checkpoint file through ```experimental_save_ceckpoint```
- Loading is done through:
  - ```experimental_load_checkpoint``` for ORTTrainer
  - pickle load for pre checkpoint state dictionary
- now the idea is to compare the state dictionary extracted through ```experimental_state_dict``` on the checkpoint loaded ORTTrainer and the state dictionary loaded through pickle.
- But these two files cannot be directly compared because one is a single node run where as the other one is a distributed run.
- So, I need to aggregate the state dictionaries for the distributed run and then compare against the originally loaded state dictionary.

These functions are the aggregation functions that are similar to the ones defined in the new aggregation logic in #5478

That is why these functions are needed. I can add docstrings to these functions, but these are temporary functions that I think will need to be removed once #5478 is merged.

Reply:
That is great information, please add this explanation on why it is needed, how you solved and any limitation that we still have after all workarounds. That helps determining how covered we are and how to debug when something unexpected happens

Reply:
Added detailed workflow comment in the main test entry point